tapalcatl_server/handler.go

```diff
@@ -163,98 +170,122 @@ type nilMetricsWriter struct{}
+	return fmt.Sprintf("%s.count:%d|c\n", makeMetricPrefix(prefix, metric), value)
+}
 
-	zipErr       metrics.Counter
-	respWriteErr metrics.Counter
-	condErr      metrics.Counter
+func makeStatsdLineGauge(prefix string, metric string, value int) string {
+	return fmt.Sprintf("%s.value:%d|g\n", makeMetricPrefix(prefix, metric), value)
 }
 
-func (smw *statsdMetricsWriter) Write(reqState *requestState) {
+func writeStatsdCount(w io.Writer, prefix string, metric string, value int) {
+	w.Write([]byte(makeStatsdLineCount(prefix, metric, value)))
+}
+
+func writeStatsdGauge(w io.Writer, prefix string, metric string, value int) {
+	w.Write([]byte(makeStatsdLineGauge(prefix, metric, value)))
+}
+
+type prefixedStatsdWriter struct {
+	prefix string
+	w      io.Writer
+}
+
+func (psw *prefixedStatsdWriter) WriteCount(metric string, value int) {
+	writeStatsdCount(psw.w, psw.prefix, metric, value)
+}
+
+func (psw *prefixedStatsdWriter) WriteGauge(metric string, value int) {
+	writeStatsdGauge(psw.w, psw.prefix, metric, value)
+}
+
+func (smw *statsdMetricsWriter) Process(reqState *requestState) {
+	conn, err := net.DialUDP("udp", nil, smw.addr)
+	if err != nil {
+		smw.logger.Printf("ERROR: Metrics Writer failed to connect to %s: %s\n", smw.addr, err)
+	}
+	defer conn.Close()
+
+	w := bufio.NewWriter(conn)
+	defer w.Flush()
+
+	psw := prefixedStatsdWriter{
+		prefix: smw.prefix,
+		w:      w,
+	}
+
 	respStateInt := int32(reqState.responseState)
 	if respStateInt > 0 && respStateInt < int32(ResponseState_Count) {
-		counter := smw.respState[respStateInt]
-		counter.Inc(1)
+		respStateName := reqState.responseState.String()
+		respMetricName := fmt.Sprintf("responsestate.%s", respStateName)
+		psw.WriteCount(respMetricName, 1)
+	} else {
+		smw.logger.Printf("ERROR: Invalid response state: %s", reqState.responseState)
 	}
 
 	fetchStateInt := int32(reqState.fetchState)
 	if fetchStateInt > 0 && fetchStateInt < int32(FetchState_Count) {
-		counter := smw.fetchState[fetchStateInt]
-		counter.Inc(1)
+		fetchStateName := reqState.fetchState.String()
+		fetchMetricName := fmt.Sprintf("fetchstate.%s", fetchStateName)
+		psw.WriteCount(fetchMetricName, 1)
+	} else {
+		smw.logger.Printf("ERROR: Invalid fetch state: %s", reqState.responseState)
 	}
 
 	if reqState.fetchSize.bodySize > 0 {
-		smw.fetchSizeBody.Update(reqState.fetchSize.bodySize)
-		smw.fetchSizeLen.Update(reqState.fetchSize.bytesLength)
-		smw.fetchSizeCap.Update(reqState.fetchSize.bytesCap)
+		psw.WriteGauge("fetchsize.body-size", int(reqState.fetchSize.bodySize))
+		psw.WriteGauge("fetchsize.buffer-length", int(reqState.fetchSize.bytesLength))
+		psw.WriteGauge("fetchsize.buffer-capacity", int(reqState.fetchSize.bytesCap))
 	}
 
-	if reqState.storageMetadata.hasLastModified {
-		smw.lastMod.Inc(1)
-	}
-	if reqState.storageMetadata.hasEtag {
-		smw.etag.Inc(1)
-	}
+	psw.WriteCount("lastmodified", boolAsInt(reqState.storageMetadata.hasLastModified))
```

Comment:
I added a new counter for all requests and sent over booleans conditionally in https://github.com/tilezen/tapalcatl/pull/13
(New pr because I had already merged this into master).